Core: Remove IE-specific support tests, rely on document.documentMode

mgol opened the pull request:
### Summary ###
<!--
Describe what this PR does. All but trivial changes (e.g. typos)
should start with an issue. Mention the issue number here.
-->

~~To make sure none of the IE-targeted workarounds depends on IE bugs we also
have a new special test run on Chrome Headless with document.documentMode
set to 11 to force all those workarounds to be applied in Chrome & make sure
tests still pass.~~

**Edit:**
1. After recent changes we're back at -412 bytes
1. Also, update some tests to IE-sniff when deciding whether to skip a test.

Fixes gh-4386

### Checklist ###
<!--
Mark an `[x]` for completed items, if you're not sure leave them unchecked and we can assist.
-->

* [x] All authors have signed the CLA at https://cla.js.foundation/jquery/jquery
* [x] New tests have been added to show the fix or feature works
* [x] Grunt build and unit tests pass locally with these changes
* ~~If needed, a docs issue/PR was created at https://github.com/jquery/api.jquery.com~~

<!--
Thanks! Bots and humans will be around shortly to check it out.
-->

mgol commented on Gruntfile.js:
```diff
@@ -163,62 +210,30 @@ module.exports = function( grunt ) {
 				browsers: [ isTravis ? "ChromeHeadlessNoSandbox" : "ChromeHeadless" ]
 			},
 
+			mainWithDocumentMode: {
```
Do you think this mode is worth it or that we should just assume IE is not going to change too much?

mgol replied:
This has been removed now.

mgol commented on test/unit/support.js:
```diff
@@ -58,53 +58,17 @@ testIframe(
 	if ( /edge\//i.test( userAgent ) ) {
 		expected = expectedMap.edge;
-	} else if ( document.documentMode ) {
+	} else if ( /trident\//i.test( userAgent ) ) {
```
I had to change this check back to be UA-based as Chrome with faked `documentMode` would still pass the support tests that IE fails.

There's no difference at the moment but if we had to add some the problem would surface.

mgol commented on test/unit/fake-document-mode.js:
```diff
@@ -0,0 +1,7 @@
+QUnit.module( "fake-document-mode", { afterEach: moduleTeardown } );
```
I haven't added this test to the list where all others are specified as it only makes sense in the mode with faked `documentMode`.

mgol commented on test/unit/support.js:
```diff
@@ -58,53 +58,17 @@ testIframe(
 	if ( /edge\//i.test( userAgent ) ) {
 		expected = expectedMap.edge;
-	} else if ( document.documentMode ) {
+	} else if ( /trident\/7\.0/i.test( userAgent ) ) {
```
I had to change this check back to be UA-based as Chrome with faked documentMode would still pass the support tests that IE fails.

There's no difference at the moment but if we had to add some the problem would surface.

mgol replied:
Reverted since we're not doing this special Chrome run.

mgol commented:
It turns out my workaround for Chrome with `documentMode` emulation failing the "SVG dimensions (border-box)" test was incorrect; IE doesn't support `offset*` properties on SVGs either. The workaround really depends on IE incorrectly reporting content-box dimensions for `border-box` elements. I reverted this change (we're down to `-415` bytes), you can see the failure on Travis now.

There are two ways we can work around this:
1. We accept that in this case we depend on the bug existing in IE and on it not getting fixed there. That means the special "Chrome with `documentMode` emulation" test run doesn't make sense anymore.
2. We restore the `boxSizingReliable` test but guard it in the additional `documentMode` check to avoid triggering layout outside of IE. This will reduce size gains from `-415` bytes to `-201` bytes, although I suspect this can be reduced by removing some now redundant CSS from the support test.

Note that in the past we've been bitten by an assumption that IE won't fix a bug when IE had a bug with dimensions in fullscreen mode inside iframes (see #1764) and we applied a workaround that multiplied the dimensions by 100 based on the `documentMode` check which we had to revert later (see #3041) when IE fixed it contrary to earlier claims. The probability of IE fixing something now is perhaps lower than 3 years ago, especially when it comes to something more complex than this simple iframe sizing bug but there's still some risk.

dmethvin commented:
I prefer the first option. It seems like the "fullscreen in iframes" bug was trivial to fix but something tells me that IE11 isn't going to fix its unreliable box sizing fix ever. If that seems like too much risk we could just guard it with `documentMode` to avoid forcing layout anywhere but IE. The sad thing about that solution is that IE11 will probably suffer the most performance hit from forcing layout.

gibson042 commented:
I also prefer the first option. As I mentioned in https://github.com/jquery/jquery/issues/4386#issuecomment-489331660 , if anyone ever _does_ release an update or fork of IE that fixes the issue, we can re-add support tests then.

mgol commented:
PR updated and is ready for final review.

I removed the special Chrome with fake docMode run, I also modified some tests to IE-sniff when deciding whether to skip instead of performing sometimes complex (& easier to do wrong) support tests where possible.

gibson042 commented on test/data/testinit.js:
```diff
@@ -256,14 +256,11 @@ if ( !window.__karma__ ) {
+// A variable to make it easier to skip specific tests in IE, mostly
+// testing integrations with newer Web features not supported by it.
+QUnit.isIE = /trident\/7\.0/i.test( window.navigator.userAgent );
+QUnit.testUnlessIE = QUnit.isIE ? "skip" : "test";
```
Should we structure this as a function rather than a method name?
```suggestion
QUnit.testUnlessIE = QUnit.isIE ?
	QUnit.skip :
	QUnit.test;
```
```patch
-QUnit[ QUnit.testUnlessIE ]( "css(--customProperty)", function( assert ) {
+QUnit.testUnlessIE( "css(--customProperty)", function( assert ) {
```

mgol replied:
Good idea, that makes much more sense.

mgol replied:
PR updated.

mgol commented:
@gibson042 PR updated. I also added a support comment for e0d941156900a6bff7c098c8ea7290528e468cf8 that was missing.

mgol commented:
@gibson042 PR updated (changes are in the new fixup commit). I also added a support comment for e0d941156900a6bff7c098c8ea7290528e468cf8 that was missing.
